sys/vm_map.c

```diff
@@ -166,16 +166,23 @@ int vm_map_findspace(vm_map_t *map, vm_addr_t start, size_t length,
     vm_map_entry_t *next = TAILQ_NEXT(it, map_list);
     if (!next)
       continue;
-    if (next->start - it->end >= length) {
-      /* We will fit into this gap. */
+    if (start < it->end && next->start - it->end >= length) {
+      /* Start points to an address before this gap, and we'll fit there. */
       *addr = it->end;
       return 0;
+    } else if (start >= it->end && next->start - start >= length) {
+      /* Start points to inside this gap, and we'll fit there. */
+      *addr = start;
+      return 0;
     }
   }
 
   /* Finally, check for free space after end. */
   vm_map_entry_t *last = TAILQ_LAST(&map->list, vm_map_list);
-  if (map->pmap->end - last->end >= length) {
+  if (start > last->end && map->pmap->end - start >= length) {
```

Comment:
I'm not sure I follow. The condition you propose does not consider the situation where `start` points inside the last mapped region and there is enough space after it to fit `length`.